lib/mkCargoDerivation.nix

```diff
@@ -39,13 +40,55 @@ let
     "pnameSuffix"
     "stdenv"
   ];
+
+  rustcWrapper = writeShellApplication {
+    name = "rustc-wrapper";
+    text = ''
+      set -euo pipefail
+
+      args=("$@")
+
+      temp_dir=$(mktemp -d)
+      trap 'rm -rf -- "$temp_dir"' EXIT
+
+      for i in "''${!args[@]}"; do
+        if [[ "''${args[i]}" == --out-dir=* ]]; then
+          current_out_dir="''${args[i]#--out-dir=}"
+          args[i]="--out-dir=$temp_dir"
+          break
+        elif [[ "''${args[i]}" == --out-dir ]]; then
+          current_out_dir="''${args[i+1]}"
+          args[i+1]="$temp_dir"
+          break
+        fi
+      done
+
+      if [[ -v current_out_dir ]]; then
+        stderr="$temp_dir/crane_stderr"
+        set +e
+        stdout=$("''${args[@]}" 2>"$stderr")
+        exit_code=$?
+        set -e
+        stderr_text=$(cat "$stderr")
+        rm "$stderr"
+        cp -r "$temp_dir/." "$current_out_dir"
+        echo -n "$stdout"
+        echo -n "$stderr_text" >&2
+        exit $exit_code
+      else
+        exec "$@"
+      fi
+    '';
+  };
 in
 chosenStdenv.mkDerivation (cleanedArgs // {
   inherit cargoArtifacts;
 
   pname = "${args.pname or crateName.pname}${args.pnameSuffix or ""}";
   version = args.version or crateName.version;
 
+  RUSTC_WRAPPER="${rustcWrapper}/bin/rustc-wrapper";
```

Comment:
I'm strongly opposed to the forced usage of `RUSTC_WRAPPER` here. Just because _we believe_ it to be rarely used is not sufficient to _outright preclude_ projects from using their own `RUSTC_WRAPPER`.

(not to mention that _plenty_ of builds work just fine without needing this workaround, so it's the exceptional builds which should opt into different behavior than forcing it on everyone)

Reply:
ack - I can handle the wrapper internally (and then also wrap with the user's wrapper).

Reply:
I'm still concerned that this would be too brittle to maintain at the library level (you can set a rustc-wrapper in `.cargo/config.toml`, and those can be littered in different places in the project).